Source/WebCore/page/DragController.cpp

```diff
@@ -239,74 +239,103 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
         clearDragCaret();
-        return true;
-    }
 
-    m_documentUnderMouse = nullptr;
-    clearDragCaret();
-
-    if (!operationForLoad(dragData))
-        return false;
+        if (!operationForLoad(dragData)) {
+            completionHandler(false);
+            return;
+        }
+        auto frameID = weakFrame->frameID();
+        client().willPerformDragDestinationAction(DragDestinationAction::Load, dragData, frameID, [weakThis = WTFMove(weakThis), completionHandler = WTFMove(completionHandler), weakFrame = WTFMove(weakFrame), shouldOpenExternalURLsPolicy, dragData] mutable {
+            RefPtr strongFrame = weakFrame.get();
+            if (!weakThis || !strongFrame) {
+                completionHandler(false);
+                return;
+            }
 
-    auto urlString = dragData.asURL();
-    if (urlString.isEmpty())
-        return false;
+            auto urlString = dragData.asURL();
+            if (urlString.isEmpty()) {
+                completionHandler(false);
+                return;
+            }
 
-    client().willPerformDragDestinationAction(DragDestinationAction::Load, dragData);
-    ResourceRequest resourceRequest { WTFMove(urlString) };
-    resourceRequest.setIsAppInitiated(false);
-    FrameLoadRequest frameLoadRequest { *localMainFrame, WTFMove(resourceRequest) };
-    frameLoadRequest.setShouldOpenExternalURLsPolicy(shouldOpenExternalURLsPolicy);
-    frameLoadRequest.setIsRequestFromClientOrUserInput();
-    localMainFrame->loader().load(WTFMove(frameLoadRequest));
-    return true;
+            ResourceRequest resourceRequest { URL { urlString } };
+            resourceRequest.setIsAppInitiated(false);
+            FrameLoadRequest frameLoadRequest { *strongFrame, WTFMove(resourceRequest) };
+            frameLoadRequest.setShouldOpenExternalURLsPolicy(shouldOpenExternalURLsPolicy);
+            frameLoadRequest.setIsRequestFromClientOrUserInput();
+            strongFrame->loader().load(WTFMove(frameLoadRequest));
+            completionHandler(true);
+        });
+    };
+    RefPtr subframe = EventHandler::subframeForTargetNode(hitTestResult.protectedTargetNode().get());
+    RefPtr remoteFrame = dynamicDowncast<RemoteFrame>(subframe);
+    if ((remoteFrame || ((m_dragDestinationActionMask.contains(DragDestinationAction::DHTML)) && dragIsHandledByDocument(m_dragHandlingMethod))) && frame.view()) {
+        if (remoteFrame) {
+            frame.eventHandler().performDragAndDrop(createMouseEvent(dragData), Pasteboard::create(dragData), dragData.draggingSourceOperationMask(), dragData.containsFiles(), hitTestResult, WTFMove(continuePerformDrag));
+            return;
+        }
+        client().willPerformDragDestinationAction(DragDestinationAction::DHTML, dragData, frame.frameID(), [dragData = WTFMove(dragData), weakFrame = WeakPtr { frame }, hitTestResult = WTFMove(hitTestResult), continuePerformDrag = WTFMove(continuePerformDrag)] mutable {
+            if (!weakFrame)
+                return;
```

Comment:
Might need to capture `RefPtr` before invoking function on frame. 
Also, `continuePerformDrag` needs to be invoked to ensure `CompletionHandler` captured in it is invoked.